### PR TITLE
Add implied bounds

### PR DESCRIPTION
**Pull request by scalexm**
These commits add implied bounds on structs and remove WF conditions for type parameters in WF predicates, e.g. before this PR we had:
```
WF(T: Clone) :- WF(T)
```
and after this PR:
```
WF(T: Clone).
```

### CODE REVIEW

File: src/lower/mod.rs [hunk 1]
```diff
@@ -1018,43 +1018,64 @@ impl Anonymize for [ir::ParameterKind<ir::Identifier>] {
...
-impl ir::StructDatum {
-    fn to_program_clauses(&self, program: &ir::Program) -> Vec<ir::ProgramClause> {
-        // Given:
-        //
-        //    struct Foo<T: Eq> { }
-        //
-        // we generate the following clause:
-        //
-        //    for<?T> WF(Foo<?T>) :- WF(?T), (?T: Eq), WF(?T: Eq).
+macro_rules! well_formed {
```
Comment (nikomatsakis) [inline src/lower/mod.rs]: It would be nice to have a comment explaining what this does =)

File: src/solve/test.rs [hunk 2]
```diff
@@ -164,36 +165,46 @@ fn prove_forall() {
...
         goal {
-            forall<T> { Vec<T>: Clone }
+            forall<T> { Vec<T>: Eq<Vec<T>> }
         } yields {
-            "CannotProve"
+            "No possible solution"
         }
 
-        // Here, we do know that `T: Clone`, so we can.
+        // Here, we do know that `T: Eq`, so we can.
```
Comment (nikomatsakis) [inline src/solve/test.rs]: Nit: `T: Eq<T>`

File: src/lower/mod.rs [hunk 3]
```diff
@@ -1018,29 +1018,46 @@ impl Anonymize for [ir::ParameterKind<ir::Identifier>] {
...
+    }
+}
+
+trait WellFormed {
+    fn wf_clauses(&self, wf_predicate: ir::WellFormed, program: &ir::Program)
+        -> Vec<ir::ProgramClause>;
+}
+
+impl<T: ExpandWhereClauses> WellFormed for ir::Binders<T> {
+    fn wf_clauses(&self, wf_predicate: ir::WellFormed, program: &ir::Program)
```
Comment (nikomatsakis) [inline src/lower/mod.rs]: I definitely prefer this to the macro, but I'd still like some comments. =)

File: src/lower/mod.rs [hunk 4]
```diff
@@ -1018,29 +1018,46 @@ impl Anonymize for [ir::ParameterKind<ir::Identifier>] {
...
-                                        .iter()
-                                        .filter_map(|pk| pk.as_ref().ty())
-                                        .cloned()
-                                        .map(|ty| ir::WellFormed::Ty(ty))
-                                        .casted();
-
-                        tys.chain(where_clauses.iter().cloned().casted()).collect()
-                    }
+                    consequence: wf_predicate.clone().cast(),
+                    conditions: where_clauses.iter().cloned().casted().collect(),
```
Comment (nikomatsakis) [inline src/lower/mod.rs]: in particular it's not obvious to me what has changed in this commit other than tests, just looking at the code anyhow

**Comment by scalexm** [comment]
Closing in favor of #82.

**Pull request closed without merging**